examples/trainer_controller/README.md

```diff
@@ -0,0 +1,122 @@
+# Trainer controller
+
+Trainer controller is a framework for controlling the trainer loop using user-defined rules and metrics.
+
+### Motivation
+
+This frameworks helps user define rules to capture scenarios like criteria for stopping an ongoing training (E.g validation loss reaching a certain target, validation loss increasing with epoch, training loss values for last 100 steps increasing etc).
+
+### Usage
+*Note: Evaluation loss and validation loss are the same.*
+1. The trainer controller feature can be used and its behavior is controlled by a configuration file (we will illustrate the configuration file below) supplied by the user at the start of the training. Here is a sample of how the user can initiate a trainer controller for a training job, by specifying path to an existing configuration `loss.yaml` in the `./examples/trainercontroller_configs` directory using the flag `--trainer_controller_config_file`:
+    ```shell
+    # if you want to use one GPU on multi-gpu machine
+    export CUDA_VISIBLE_DEVICES=0
+
+    # MODEL_PATH=meta-llama/Llama-2-7b-hf # Huggingface model id or path to a checkpoint
+    # TRAIN_DATA_PATH=twitter_complaints.json # Path to the dataset
+                    # contains data in form of [{"input": text , "output": text}]
+    # VALIDATION_DATA_PATH=/path/to/validation/dataset
+    # OUTPUT_PATH=out # Path to the output folder where the checkpoints are saved
+
+    # EXAMPLE_CONFIGS="./examples/trainercontroller_configs"
+
+    python ./tuning/sft_trainer.py  \
+    --model_name_or_path $MODEL_PATH  \
+    --tokenizer_name_or_path $MODEL_PATH \
+    --training_data_path $TRAIN_DATA_PATH  \
+    --output_dir $OUTPUT_PATH  \
+    --validation_data_path $VALIDATION_DATA_PATH \
+    --num_train_epochs 10  \
+    --per_device_train_batch_size 4  \
+    --per_device_eval_batch_size 4  \
+    --gradient_accumulation_steps 4  \
+    --evaluation_strategy "epoch"  \
+    --save_strategy "epoch"  \
+    --learning_rate 1e-5  \
+    --weight_decay 0.  \
+    --warmup_ratio 0.03  \
+    --lr_scheduler_type "cosine"  \
+    --logging_steps 5  \
+    --logging_strategy "steps" \
+    --include_tokens_per_second  \
+    --packing False  \
+    --metric_for_best_model "loss" \
+    --load_best_model_at_end True \
+    --use_flash_attn False \
+    --trainer_controller_config_file $EXAMPLE_CONFIGS"/loss.yaml" \
+    --response_template "\n### Response:"  \
+    --dataset_text_field "output
```

Comment:
We can highlight only the controller config flag

Reply:
@ashokponkumar Markdown does not allow for highlighting in code block.